primitives.lisp

```diff
@@ -561,18 +562,131 @@ Use the window's resource class.
+  "Make an instance of a StumpWM class and autoenable any relevant minor
+modes. CLASS must be a symbol denoting a class which descends, directly or
+indirectly, from swm-class. INITARGS must be all initargs one would pass to
+make-instance."
+  ;; This is implemented as a function instead of as an after method for
+  ;; initialize-instance because autoenabling a minor mode involves changing the
+  ;; class of the object, which is implied to be undefined behavior if called
+  ;; within a method which accesses the objects slots.
+  (declare (special *active-global-minor-modes*))
+  (let ((object (apply #'make-instance class initargs)))
+    (prog1 object
+      (loop for class in *active-global-minor-modes*
+            when (typep object (scope-type (minor-mode-scope class)))
+              do (autoenable-minor-mode class object))
+      (setf (swm-class-new-objects object)
+            (remove object (swm-class-new-objects object) :test #'eq)))))
+
+(defmacro define-swm-class (class-name superclasses slots &rest options)
+  "Define a class and a method for DYNAMIC-MIXINS:REPLACE-CLASS which specializes
+upon the class and replaces it. If SUPERCLASSES is NIL then (SWM-CLASS) is used."
+  (unless superclasses (setq superclasses '(swm-class)))
+  `(progn
+     (defclass ,class-name ,superclasses ,slots ,@options)
+     (defmethod dynamic-mixins:replace-class ((object ,class-name) new &rest r)
+       (apply #'dynamic-mixins:replace-class-in-mixin
+              object new ',class-name r))))
+
+(define-swm-class frame ()
+  ((number
+    :initform nil
+    :initarg :number
+    :accessor frame-number)
+   (x
+    :initform nil
+    :accessor frame-x
+    :initarg :x)
+   (y
+    :initform nil
+    :accessor frame-y
+    :initarg :y)
+   (width
+    :initform nil
+    :accessor frame-width
+    :initarg :width)
+   (height
+    :initform nil
+    :accessor frame-height
+    :initarg :height)
+   (window
+    :initform nil
+    :accessor frame-window
+    :initarg :window)))
+
+(defun frame-p (object)
+  (typep object 'frame))
+
+(defun make-frame (&rest rest &key number x y width height window)
+  (declare (ignore number x y width height window))
+  (apply 'make-swm-class-instance 'frame rest))
+
+(defun copy-frame (instance)
+  (make-swm-class-instance 'frame :number (frame-number instance)
+                                  :x (frame-x instance)
+                                  :y (frame-y instance)
+                                  :width (frame-width instance)
+                                  :height (frame-height instance)
+                                  :window (frame-window instance)))
+
+(define-swm-class head (frame)
+  ((name
+    :initform ""
+    :accessor head-name
+    :initarg :name)))
+
+;; duplicate frame accessors for heads.
+(macrolet ((define-head-accessor (name)
+             (let ((pkg (find-package :stumpwm)))
+               `(progn
+                  (defgeneric ,(intern (format nil "HEAD-~A" name) pkg) (head)
```

Comment:
```suggestion
                  (defgeneric ,(sb-int:package-symbolicate pkg "HEAD-" name)
```

consider using package-symbolicate, although idk how the SBCL people would feel abou using their internal utility package ourselves :thinking:

Reply:
Im not familiar with package-symbolicate, but I'll take a look. I'm hesitant to use internals given that it may change without warning and across versions, and some folks might use a newer/older version of SBCL.